### PR TITLE
Brittany plugin and support for textDocument/formatting, textDocument/rangeFormatting

### LINKED ISSUE #215: Integrate brittany
See https://github.com/lspitzner/brittany/issues/5

### PR DESCRIPTION
**Pull request by wz1000**
Range formatting doesn't take into account pragmas yet, so often fails.

~~Also, we don't take into account client supplied Formatting Options yet.~~

Closes #215 and lspitzner/brittany#5

### CODE REVIEW

File: stack.yaml [hunk 1]
```diff
@@ -34,18 +35,28 @@ packages:
...
 - pipes-aeson-0.4.1.8
 - pipes-network-0.6.4.1
+- butcher-1.1.0.0
+- czipwith-1.0.0.0
+- data-tree-print-0.1.0.0
+- deque-0.2
+- monad-memo-0.4.1
 # - hspec-jenkins-0.1.1
 flags:
   haskell-ide-engine:
```
Comment (alanz) [inline stack.yaml]: please add a pedantic flag down here for `hie-brittany` too. And we need to move the CI ti run in pedantic too.

Comment (wz1000) [inline stack.yaml]: ok

File: hie-brittany/Haskell/Ide/BrittanyPlugin.hs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import qualified GHC.LanguageExtensions.Type as GHC
+import qualified Data.Text.Lazy as TextL
+import qualified GhcMod.Utils as GM
+import qualified Language.Haskell.LSP.TH.DataTypesJSON as J
+import           Control.Lens
+
+
+brittanyCmd :: TextDocumentIdentifier -> Maybe Range -> IdeM (IdeResponse [J.TextEdit])
+brittanyCmd tdi range =
+  pluginGetFile "genapplicative: " (tdi ^. J.uri) $ \file -> do
```
Comment (alanz) [inline hie-brittany/Haskell/Ide/BrittanyPlugin.hs]: There are a number of strings for `pluginGetFile` that do not actually match. Not sure how we can test for them though. Perhaps better to leave them out and let the exception location guide us?

**Comment by alanz** [comment]
Made a few small comments. Please also refer to closing https://github.com/haskell/haskell-ide-engine/issues/215 in the commit message.

File: hie-brittany/Haskell/Ide/BrittanyPlugin.hs (hunk 2 above)
Comment (wz1000) [inline hie-brittany/Haskell/Ide/BrittanyPlugin.hs]: I'll go over all of them once.

**Pull request merged**